src/hotspot/share/opto/compile.cpp

```diff
@@ -3677,13 +3677,18 @@ void Compile::final_graph_reshaping_impl(Node *n, Final_Reshape_Counts& frc, Uni
 
 void Compile::final_graph_reshaping_main_switch(Node* n, Final_Reshape_Counts& frc, uint nop, Unique_Node_List& dead_nodes) {
   switch( nop ) {
-  // Count all float operations that may use FPU
+  // Count all half-precision float operations that may use the FPU
   case Op_AddHF:
   case Op_SubHF:
   case Op_MulHF:
   case Op_DivHF:
   case Op_MaxHF:
   case Op_MinHF:
+  case Op_AbsHF:
+  case Op_NegHF:
+  case Op_SqrtHF:
+  case Op_FmaHF:
+  // Count all float operations that may use FPU
```

Comment:
Hey @Bhavana-Kilambi , I am little worried about introducing specialized IR nodes as discussed in [pervious thread](https://github.com/openjdk/valhalla/pull/1196#issuecomment-2283627677),  even though our current strategy fits well with C2 compilation flow without causing much disturbance, going forward we may need to support other reduced precision types BF16 (AVX10.2 has several ISA extensions for BF16), FP8 and INT8, creating unique IR may not be desirable, lets revisit this after features completion with conversion ops.

Reply:
I have the conversion ops and isFinite/Infinite/NaN intrinsics still left to complete.

Reply:
I am in process of porting Float16 extension to VectorAPI into lworld+fp16, that will consolidate Flot16 effort on lworld+fp16 and leverage the unified backed implementation. With VectorAPI we will also need to support Reductions for Float16 vectors :-)

Reply:
@jatin-bhateja Thanks for the effort. It will be good to test Float16 with vectorAPI and add the missing operations. I think we need to support Float16 reductions in auto-vectorization as well? I dont think those operations are added for FP16. 
Also about the design part, I was thinking instead of modifying the existing opcode or adding a separate secondary opcode for the operations, can we reuse the same existing FP32 nodes for FP16 as well? We can add flags for the FP32 IR nodes to check if it is a FP16 op or eventually FP8 op ? That way we can reuse the  same Value(), Ideal(), identity() methods and can modify those functions based on the value these flags returns in case we do not want a particular optimization to be applied for FP16/FP8? In the backend, we can use these flags to emit the correct instruction with the correct SIMD arrangement ? I guess this might introduce a lot of checks (to check value of the flags) in the compiler code.. What are your thoughts about it?

Reply:
Also I will address review comments and post a new patch most likely after Aug 26th. I'll be on annual leave (plus UK bank holiday) from Aug 22nd to Aug 26th.

Reply:
> @jatin-bhateja Thanks for the effort. It will be good to test Float16 with vectorAPI and add the missing operations. I think we need to support Float16 reductions in auto-vectorization as well? I dont think those operations are added for FP16. Also about the design part, I was thinking instead of modifying the existing opcode or adding a separate secondary opcode for the operations, can we reuse the same existing FP32 nodes for FP16 as well? We can add flags for the FP32 IR nodes to check if it is a FP16 op or eventually FP8 op ? 

With new flag in existing IR approach we may end up adding complimentary predication checks in existing floating point instruction selection patterns.

> That way we can reuse the same Value(), Ideal(), identity() methods and can modify those functions based on the value these flags returns in case we do not want a particular optimization to be applied for FP16/FP8? 

Yes, idea is to reuse existing idealization routines, either by factoring out the meaty parts, or by inheritance. 

> In the backend, we can use these flags to emit the correct instruction with the correct SIMD arrangement ? I guess this might introduce a lot of checks (to check value of the flags) in the compiler code.. What are your thoughts about it?

I agree.
 
To me having a secondary opcode along with some code re-factoring around value / ideal / identity routines look cleaner, but we can think over alternatives which fits best in existing infrastructure.